### PR TITLE
fix: event read from syslog when syslog entry too long

### PR DESCRIPTION
**Pull request by matejvasek**
When labels map is too big we may get syslog entry truncated.
This breaks JSON parsing making event loading impossible.

```
ERRO[0016] Unable to decode event: unexpected end of JSON input
```

By default the limit is 64KiB, this PR set it to 0 == unlimited.

```release-note
fix: increased event syslog deserialization buffer 
```

related to:
* https://github.com/containers/podman/issues/16834
* https://github.com/containers/container-selinux/pull/196

### CODE REVIEW

**Comment by matejvasek** [comment]
This is imo worth backporting to 4.3.

**Comment by rhatdan** [comment]
@giuseppe @Luap99 @mheon PTAL
I wish there was a way to figure out the max, rather then hard coding it.

**Comment by matejvasek** [comment]
> @giuseppe @Luap99 @mheon PTAL
I wish there was a way to figure out the max, rather then hard coding it.

640K ought to be enough for anybody.

**Comment by matejvasek** [comment]
My specific case needs `104059` bytes.

**Comment by Luap99** [comment]
From `sd_journal_get_data(3)`

> sd_journal_set_data_threshold() may be used to change the data field size threshold for data returned by sd_journal_get_data(), sd_journal_enumerate_data() and sd_journal_enumerate_unique(). This threshold is a hint only: it indicates that the client program is interested only in the initial parts of the data fields, up to the threshold in size — but the library might still return larger data objects. That means applications should not rely exclusively on this setting to limit the size of the data fields returned, but need to apply an explicit size limit on the returned data as well. This threshold defaults to 64K by default. To retrieve the complete data fields this threshold should be turned off by setting it to 0, so that the library always returns the complete data objects. It is recommended to set this threshold as low as possible since this relieves the library from having to decompress large compressed data objects in full.

Given that we always need to full information we should set it to `0` instead, at least according to this description I have not tested it.

@matejvasek You should also add a test with labels > 64 KiB to make sure we do not regress again.

**Comment by matejvasek** [comment]
@Luap99 I did add a test at first, but it was failing in CI for some reason so I reverted it.

**Comment by matejvasek** [comment]
The test worked locally perfectly fine, and I cannot debug what is wrong in CI.

**Comment by matejvasek** [comment]
I set the threshold to 0 as suggested @Luap99 .

**Comment by matejvasek** [comment]
@giuseppe @Luap99 @mheon 
Any idea why test failing (timing out) in CI? I can only think of `events_logger="none"`, but I doubt that is used there.

**Comment by Luap99** [comment]
I think API test force `--events-backend file`. I suggest you create a new test in test/system/090-events.bats, I it is enough to create create a container with a big label and then podman events should fail right now?

**Comment by matejvasek** [comment]
@Luap99 but the `file` backend should work too, although it wouldn't test the actuall issue with journald.

**Comment by Luap99** [comment]
> @Luap99 but the `file` backend should work too, although it wouldn't test the actuall issue with journald. Only the `none` backend should timeout.

Yeah I am not sure what the API tests use, I recommend to use the system tests and run this for both file and journald driver. This makes it more explicit what is actually being tested.

**Comment by mheon** [comment]
Code changes LGTM

**Comment by matejvasek** [comment]
@Luap99 I am close to finishing tests in `test/system/090-events.bats`.
It has however one tiny issue, the events sub-command or testing framework cannot handle big labes :rofl: .

```
# Error: unable to process labels: bufio.Scanner: token too long
```

**Comment by matejvasek** [comment]
@Luap99 @mheon PTAL

File: test/system/090-events.bats [hunk 1]
```diff
@@ -291,3 +291,21 @@ EOF
...
+@test "event with too big labels" {
+    # test that events entries are not truncated when reading from syslog
+
+    cname=test-$(random_string 30 | tr A-Z a-z)
+
+    local label_file="$BATS_TEST_TMPDIR/labels.txt"
+    echo -n "too-long-label=" >> "$label_file"
+    random_string 150000 >> "$label_file"
+
+    local start="$(date --iso-8601=ns)"
```
Comment (matejvasek) [inline test/system/090-events.bats:304]: The man page says that rfc-3339 (2022-12-08 18:23:31.758649447+01:00) should be used,
but that doesn't work.
The iso format works (2022-12-08T18:24:25,699552059+01:00).
Why?

**Comment by matejvasek** [comment]
> @Luap99 I am close to finishing tests in `test/system/090-events.bats`. It has however one tiny issue, the podman CLI and/or testing framework cannot handle big labes rofl .
> 
> ```
> # Error: unable to process labels: bufio.Scanner: token too long
> ```

I had to increase scanner buffer in parser of `label-file`.

File: test/system/helpers.bash [hunk 2]
```diff
@@ -747,7 +747,7 @@ function parse_table() {
 function random_string() {
     local length=${1:-10}
 
-    head /dev/urandom | tr -dc a-zA-Z0-9 | head -c$length
+    tr -dc a-zA-Z0-9 < /dev/urandom | head -c$length
```
Comment (rhatdan) [inline test/system/helpers.bash:750]: @edsantiago PTAL

Comment (matejvasek) [reply test/system/helpers.bash:750]: THB I don't know what that head does there. Why it was there?

Head takes 10 lines from a file, that makes as sense for text files not /dev/urandom.
For all we know first 10 chars in /dev/urandom may be new lines (I know, it is really unlikely) meaning `random_string` would return empty string.

`random_string 1000` is returning random length string not strings 1000 chars long.

Comment (matejvasek) [reply test/system/helpers.bash:750]: old verison:
```
for i in $(seq 10); do random_string 1000 | wc -c; done
686
468
512
735
288
954
749
458
463
891
```

File: test/system/090-events.bats (hunk 1 above)
Comment (matejvasek) [reply test/system/090-events.bats]: Either Go, or the `date` util have got some formats wrong.

Comment (matejvasek) [reply test/system/090-events.bats]: Probably GNU date bug?

File: test/system/helpers.bash (hunk 2 above)
Comment (edsantiago) [reply test/system/helpers.bash:750]: The first `head` was there because it's Good Enough™.

File: test/system/090-events.bats (hunk 1 above)
Comment (edsantiago) [reply test/system/090-events.bats]: Sorry, what's the question? If the man page says 4rfc3339, my first guess would be that the person who wrote it wasn't 100% clear on the differences between rfc3339 and iso8601. (Which is fine, because I've been dealing with time representations for thirty years and still have trouble with it.) Since iso8601 works, I'd suggest fixing the man page.

Comment (matejvasek) [reply test/system/090-events.bats]: The issues is that GNU `date` util separates date & time by a space, not by capital T.
The iso8601 format uses `T` as it should but it uses comma `,` as decimal separator so it is wrong too.
I have given up and now I am using `--since=10s`.

Comment (edsantiago) [reply test/system/090-events.bats]: That's unfortunate, and it's very clearly a bug: its own documentation states that it uses a period. Seems worth reporting to bug-coreutils@gnu.org.

Comment (matejvasek) [reply test/system/090-events.bats]: from wiki
>A decimal mark, either a comma or a dot is used as a separator between the time element and its fraction. (Following ISO 80000-1 according to ISO 8601:1-2019,[26] it does not stipulate a preference except within International Standards, but with a preference for a comma according to ISO 8601:2004.

File: test/system/090-events.bats [hunk 3]
```diff
@@ -291,3 +291,22 @@ EOF
...
 }
+
+@test "event with too big labels" {
+    # test that events entries are not truncated when reading from syslog
+
+    skip_if_remote "remote does not support --events-backend"
+
+    cname=test-$(random_string 30 | tr A-Z a-z)
+
+    local label_file="$BATS_TEST_TMPDIR/labels.txt"
```
Comment (edsantiago) [inline test/system/090-events.bats:302]: No such thing. You want `PODMAN_TMPDIR` https://github.com/containers/podman/blob/6e2e9ab2273f0ec6d2a993b8968b34b045509e02/test/system/helpers.bash#L81-L86

File: test/system/090-events.bats [hunk 4]
```diff
@@ -291,3 +291,22 @@ EOF
...
+    echo -n "too-long-label=" >> "$label_file"
+    random_string 150000 >> "$label_file"
```
Comment (edsantiago) [inline test/system/090-events.bats:304]: Maybe it's my old age, but I get squicked out by things like this, because my gut screams that it will block waiting for entropy. Maybe that's not a problem these days? Anyhow, I see little reason to find out. This doesn't really have to be a random string. But I won't block over it.

File: test/system/090-events.bats [hunk 5]
```diff
@@ -291,3 +291,22 @@ EOF
     _events_container_create_inspect_data journald
     _events_container_create_inspect_data file
 }
+
+@test "event with too big labels" {
+    # test that events entries are not truncated when reading from syslog
+
+    skip_if_remote "remote does not support --events-backend"
```
Comment (edsantiago) [inline test/system/090-events.bats:298]: You almost certainly also want `skip_if_journald_unavailable`

**Review by edsantiago — COMMENTED** [review]
Rootless tests are going to fail because of the invalid tmpdir. When you repush, please squash your commits.

Thanks for your work on this! Nice corner case.

**Comment by edsantiago** [comment]
And, whew, CI failed. It's not just me. The `start` event is never actually getting logged. It's not throwing an error, but it's not getting logged. Actually, **no** events for this container are being logged.

File: test/system/090-events.bats (hunk 3 above)
Comment (matejvasek) [reply test/system/090-events.bats:302]: You sure? Fore me it returns more than just `/tmp/`.
e.g 
`/tmp/bats-run-m6qynu/test/1/`

Comment (matejvasek) [reply test/system/090-events.bats:302]: it should be specific for each test case

File: test/system/090-events.bats (hunk 4 above)
Comment (matejvasek) [reply test/system/090-events.bats:304]: It would block on `/dev/random` but not `/dev/urandom` I think.

File: test/system/090-events.bats (hunk 3 above)
Comment (edsantiago) [reply test/system/090-events.bats:302]: Oh, nice. TIL about `$BATS_TEST_TMPDIR`. But it's new in bats 1.4, and RHEL still uses 1.0 or 1.1, so yeah, sorry, you need to use `$PODMAN_TMPDIR`. It is unique to each `@test`.

**Comment by matejvasek** [comment]
@edsantiago suggested changes incorporated.

File: test/system/090-events.bats (hunk 5 above)
Comment (matejvasek) [reply test/system/090-events.bats:298]: fixed

File: test/system/090-events.bats (hunk 4 above)
Comment (matejvasek) [reply test/system/090-events.bats]: fixed

File: test/system/090-events.bats (hunk 3 above)
Comment (matejvasek) [reply test/system/090-events.bats]: fixed

File: test/system/090-events.bats (hunk 4 above)
Comment (matejvasek) [reply test/system/090-events.bats]: I just just Good Enough™ approach with loop. Is there more more efficient way?

**Comment by edsantiago** [comment]
Thanks - but the code doesn't actually, like, work. As mentioned above, the events are not logged. It is super trivial to confirm this:
```console
terminal1$ bin/podman --events-backend=journald events
```
```console
terminal2$ (echo -n foo=;seq 1 50000|tr -cd 0-9) >| /tmp/mylabel 
terminal2$ bin/podman --events-backend=journald run --name foo --label-file=/tmp/mylabel --rm quay.io/libpod/testimage:20221018 date
```
In case it matters: 6.0.9-300.fc37.x86_64 systemd-251.8-586.fc37.x86_64

**Comment by edsantiago** [comment]
My week is almost over (I work 4-day weeks) and I'm on PTO the next two weeks. I will check in as time allows this weekend, but that may not be much nor will it be real-time. Good luck, and thanks again. (And don't forget to squash your commits before submitting for final review).

**Comment by edsantiago** [comment]
Quick quick followup: the discrepancy is an AVC. On my system, and presumably in CI as well, the long label triggers an AVC:
```
time->Thu Dec  8 15:37:24 2022
type=PROCTITLE msg=audit(1670539044.227:45451): proctitle="/usr/lib/systemd/systemd-journald"
type=SOCKADDR msg=audit(1670539044.227:45451): saddr=0100003730653963
type=SYSCALL msg=audit(1670539044.227:45451): arch=c000003e syscall=47 success=yes exit=0 a0=4 a1=7ffcb9bf5ec0 a2=40000040 a3=7ffcb9bf9080 items=0 ppid=1 pid=3613 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="systemd-journal" exe="/usr/lib/systemd/systemd-journald" subj=system_u:system_r:syslogd_t:s0 key=(null)
type=AVC msg=audit(1670539044.227:45451): avc:  denied  { read write } for  pid=3613 comm="systemd-journal" path=2F6465762F73686D2F6A6F75726E616C2E5858585858333432343638333132202864656C6574656429 dev="tmpfs" ino=70516 scontext=system_u:system_r:syslogd_t:s0 tcontext=unconfined_u:object_r:container_runtime_tmpfs_t:s0 tclass=file permissive=0
```
On @matejvasek's system, it does not. 

On my system, with `setenforce 0`, the log entries make it into the journal.  So does `s/50000/23445/` in my `seq` reproducer above. It is a line-length issue. With `23446`, it fails.

I'm out of time today, sorry.

**Comment by matejvasek** [comment]
@rhatdan ^^^ any ideas about the SELinux issues?

**Comment by matejvasek** [comment]
>On @matejvasek's system, it does not.

Also note I have SELinux enabled on Fedora 36.

**Comment by matejvasek** [comment]
Also why do we log labels?

**Comment by edsantiago** [comment]
More info: it only fails if the `podman` binary is `container_runtime_exec_t`. That's almost certainly why it doesn't fail for you, @matejvasek . Try this:
```console
$ chcon -t container_runtime_exec_t bin/podman
```
...and then try my reproducers, and then try them with `23445` (works) and `23446` (fails).

I really, really have to leave now.

**Comment by mheon** [comment]
@matejvasek I think it's for filtering events, but I'm not 100% sure about that.

**Comment by matejvasek** [comment]
@edsantiago I looks like you are correct, after chcon the test fails.

**Comment by edsantiago** [comment]
Okay, here's a compromise. @matejvasek write your test like this:
```bash
    (echo -n l=;head -c106122 /dev/zero|tr '\0' 'x') >| /tmp/mylabel
    $PODMAN --events-backend=journald run --name foo --label-file=/tmp/mylabel --pull=never --rm $IMAGE date  
```
(or something along those lines, use random chars if you like). That will FAIL with podman@main, as expected:
```
Error: unable to process labels: bufio.Scanner: token too long
```
...but it will PASS with podman @ your-pr. 106122 is the magic cutoff: one more byte, and it semireliably fails with the AVC (not always) (and no, I'm not going to investigate why the not-always). To play it safe, trim off a few more bytes. That should give you a pretty decent test of your new functionality.

**Comment by matejvasek** [comment]
But 106122 is dangerously close to what we actually have.

**Comment by matejvasek** [comment]
We actually use image that has 100K long labels.

**Comment by matejvasek** [comment]
```
docker image inspect gcr.io/paketo-buildpacks/builder:base | jq -r '.[0].Config.Labels'  | wc -c
104282
```

**Comment by matejvasek** [comment]
@rhatdan any idea why binary with `container_runtime_exec_t` cannot write long journal entries?

**Comment by rhatdan** [comment]
It is now allowed in container-selinux.
https://github.com/containers/container-selinux/releases/tag/v2.194.0

**Comment by matejvasek** [comment]
@rhatdan  Thanks a lot for help.

Out of curiosity, why it happens only for long entries? Buffering or something else?
How did you managed to debugged the issue?

I assume it takes some time to propagate selinux policies to Linux distributions.
Would you agree with merging this PR without tests for now?

File: cmd/podman/parse/net.go [hunk 6]
```diff
@@ -138,6 +138,8 @@ func parseEnvOrLabelFile(envOrLabel map[string]string, filename, configType stri
 	defer fh.Close()
 
 	scanner := bufio.NewScanner(fh)
+	buff := make([]byte, 256*(1<<10))
+	scanner.Buffer(buff, len(buff))
```
Comment (Luap99) [inline cmd/podman/parse/net.go:142]: This limit feels arbitrary. Can you document why this is chosen? I feel the correct way would be to stat the file and alloc the actual size?

**Comment by Luap99** [comment]
I am fine with skipping the test if you create an issue for it so we do not forget to unskip it once the fix is the VM images.

File: cmd/podman/parse/net.go (hunk 6 above)
Comment (matejvasek) [reply cmd/podman/parse/net.go:142]: Yes, it is totally arbitrary.
TBH I updated this code only for testing sake. In reality long labels are not created via `--labels-file`, but by using image which has big labels.

**Comment by matejvasek** [comment]
@Luap99 @mheon I removed tests (and created an issue for their future addition) PTAL

**Comment by Luap99** [comment]
> @Luap99 @mheon I removed tests (and created an issue for their future addition) PTAL

That is not what I meant, I would like you to keep it and just add a `skip "Issue #XXXX: selinux problem with to long labels"` this makes it easer for someone to later enable it and also keeps the git history more accurate.

**Comment by matejvasek** [comment]
> > @Luap99 @mheon I removed tests (and created an issue for their future addition) PTAL
> 
> That is not what I meant, I would like you to keep it and just add a `skip "Issue #XXXX: selinux problem with to long labels"` this makes it easer for someone to later enable it and also keeps the git history more accurate.

done

File: test/python/docker/compat/test_containers.py [hunk 7]
```diff
@@ -260,3 +262,24 @@ def test_mount_rw_by_default(self):
                 ctr.remove()
             if vol is not None:
                 vol.remove(force=True)
+
+    def test_wait_next_exit(self):
+        self.skipTest("Skip until fix container-selinux#196 is available.")
```
Comment (matejvasek) [inline test/python/docker/compat/test_containers.py:267]: Alternatively I could try to detect current version of `container-selinux` pkg and skip test conditionally.
I am not sure if it easy/worth.

Comment (matejvasek) [reply test/python/docker/compat/test_containers.py:267]: fix should be in `v2.194.0`

**Comment by rhatdan** [comment]
As far as why this is happening, in  the first place. I think the kernel is creating a temporary file in tmpfs when the buffer is big enough.  This file is labeled from an SELinux point with a tmpfs_t label. Since this has not happened in the past, the policy did not allow it by default. Bottom line I believe this is normal behavior.

**Comment by matejvasek** [comment]
@Luap99 @mheon are latest test failures flakes? I don't see how they are related to my changes.

File: test/python/docker/compat/test_containers.py (hunk 7 above)
Comment (vrothberg) [reply test/python/docker/compat/test_containers.py:267]: @cevich could we update the CI images?

**Comment by rhatdan** [comment]
LGTM

File: test/python/docker/compat/test_containers.py (hunk 7 above)
Comment (cevich) [reply test/python/docker/compat/test_containers.py:267]: Image update coming in https://github.com/containers/podman/pull/16692

Not sure if it includes any desired fixes but given a version number, I can check.

**Review by giuseppe — APPROVED** [review]
LGTM

File: test/python/docker/compat/test_containers.py (hunk 7 above)
Comment (edsantiago) [reply test/python/docker/compat/test_containers.py:267]: Yuk. I'm sorry this became pythonized: (1) it will never get unskipped, because my fixme-check script doesn't check python, and (2) it will never get run anyway because gating tests only run the BATS tests, not python. @matejvasek could I ask you to submit a new PR with your system test as it was before?

**Comment by matejvasek** [comment]
@rhatdan @mheon will be this backported to 4.3?

**Comment by mheon** [comment]
Probably not, we're starting to work on 4.4 for a late January release now.

**Comment by matejvasek** [comment]
> Probably not, we're starting to work on 4.4 for a late January release now.

@mheon 
Will Fedora use it soon? What about RHEL?

**Comment by mheon** [comment]
For Fedora, it should land as soon as it releases (minus delay waiting for the bodhi updates to make it to stable). For RHEL, RHEL 8.8 will get it.

**Comment by matejvasek** [comment]
@mheon 
I am asking because this is affect not only `podman events` but also Docker compat. API.
Things might not work for people using podman as docker daemon replacement.

**Comment by mheon** [comment]
How so? Would this be considered a breaking change?

**Comment by matejvasek** [comment]
@mheon 
The `wait` endpoint must be able to detect transition between running->exited.
In order to do that we use events (I implemented that, polling is not reliable).

Without this PR the events are not written and/or read.
For example this causes freeze for [`pack`](https://github.com/buildpacks/pack) CLI.

**Comment by matejvasek** [comment]
We noticed problems after https://github.com/containers/podman/pull/15633 (4.3).
In that PR we started writing labels for start and exit events.

**Comment by mheon** [comment]
Ack, OK. I don't think we're considering another 4.3 release, but if we do this should definitely be in consideration.

**Comment by rhatdan** [comment]
Yes no 4.3 updates, you will get it in 4.4.  We should have RC1 available this week.

**Comment by matejvasek** [comment]
@rhatdan  `RHEL 8.7` seems to use `4.2` and `RHEL 8.8` will use `4.4` so we are good I guess?

**Comment by matejvasek** [comment]
Or will any `RHEL` use 4.3?

**Comment by matejvasek** [comment]
I noticed `v4.3.1-rhel` branch exists, how it is used?

**Comment by rhatdan** [comment]
It is not used, 4.3 was never released to RHEL, this branch was created in anticipation of releasing it to RHEL, but since the releases fell in the wrong way we decided to skip that release.  RHEL8.8 and 9.2 will get 4.4

**Comment by mheon** [comment]
4.3.1-rhel was used for a CentOS stream release or two, I think. It won't make it into RHEL proper.

**Comment by matejvasek** [comment]
@rhatdan @mheon thanks, that's sounds good!

What about the SELinux policy pkg `container-selinux v2.198.0` will it available in RHEL and Fedora?

**Comment by matejvasek** [comment]
@rhatdan @mheon fedora-coreos used by podman machine still uses 4.3.1.
It is an issue for people on macOS.
Maybe we should have backported after all? Or will fedora-coreos receive 4.4 soon?
